packages/fuels-types/src/bech32.rs

```diff
@@ -0,0 +1,154 @@
+use crate::errors::Error;
+use bech32::Variant::Bech32m;
+use bech32::{FromBase32, ToBase32};
+use core::str;
+use fuel_tx::{Address, Bytes32, ContractId};
+use std::fmt::{Display, Formatter};
+
+pub const FUEL_BECH32_HRP: &str = "fuel";
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum ContentType {
+    Address,
+    ContractId,
+}
+
+/// An address in the bech32 format consisting of a
+/// human-readable part (HRP), '1' as a separator, the
+///  public key hash encoded with a u5 charset, and a checksum.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct Bech32 {
+    /// Public key hash or contract hash
+    hash: Bytes32,
+
+    /// String representation of the bech32 format
+    encoding: String,
+
+    pub content_type: ContentType,
+}
+
+impl Bech32 {
+    pub fn new_address(hrp: &str, pubkey_hash: [u8; 32]) -> Bech32 {
+        Self::new(hrp, pubkey_hash, ContentType::Address)
+    }
+
+    pub fn new_contract_id(hrp: &str, contract_hash: [u8; 32]) -> Bech32 {
+        Self::new(hrp, contract_hash, ContentType::ContractId)
+    }
+
+    fn new(hrp: &str, data: [u8; 32], content_type: ContentType) -> Self {
+        let data_base32 = data.to_base32();
+        let encoding = bech32::encode(hrp, &data_base32, Bech32m).unwrap();
+
+        Self {
+            hash: Bytes32::from(data),
+            encoding,
+            content_type,
+        }
+    }
+
+    pub fn to_address(&self) -> Address {
+        Address::new(*self.hash)
+    }
+
+    pub fn to_contract_id(&self) -> ContractId {
+        ContractId::new(*self.hash)
+    }
+
+    pub fn new_address_from_string(s: &str) -> Result<Self, Error> {
+        Self::from_string(s, ContentType::Address)
+    }
+
+    pub fn new_contract_id_from_string(s: &str) -> Result<Self, Error> {
+        Self::from_string(s, ContentType::ContractId)
+    }
+
+    fn from_string(s: &str, content_type: ContentType) -> Result<Self, Error> {
+        let (_, pubkey_hash_base32, _) = bech32::decode(s)?;
+
+        let pubkey_hash: [u8; Address::LEN] = Vec::<u8>::from_base32(&pubkey_hash_base32)?
+            .as_slice()
+            .try_into()?;
+
+        Ok(Self {
+            hash: Bytes32::from(pubkey_hash),
+            encoding: s.to_string(),
+            content_type,
+        })
+    }
+}
+
+#[allow(clippy::from_over_into)]
+impl Into<Address> for &Bech32 {
```

Comment:
Why not go for:
```rust
impl From<&Bech32> for Address { ...
```

and make clippy happy?

Reply:
That impl would need to go upstream to `fuel-types` and it seemed odd to have `Bech32` in there since it is only used on the SDK side.

Reply:
rust doesn't seem to mind having it here? just back from compiling it. can you re-check, maybe I'm doing something wrong?